File: packages/julia/src/Seguranca/Ataques/AtaqueChaveErrada.jl
"""
Garante que um token só decifra com a chave correta: com qualquer outra chave de
32 bytes, decrypt() tem que lançar.

Cuidado: CriptografiaAlvo escreve a chave no ambiente do processo; a chave
original é restaurada no final pra não afetar os outros ataques.
"""
struct AtaqueChaveErrada <: Ataque
    tentativas::Int
end

AtaqueChaveErrada(; tentativas::Integer = 30) = AtaqueChaveErrada(Int(tentativas))

nome(::AtaqueChaveErrada) = "Rejeição de chave incorreta"

function executar(a::AtaqueChaveErrada, alvo::AlvoCriptografico)
    if !(alvo isa CriptografiaAlvo)
        throw(SkipAtaqueException("Precisa de CriptografiaAlvo para trocar a chave."))
    end

    chave_original = chave_de_teste(alvo)
    tokens = [encrypt(alvo, "MENSAGEM_SECRETA_$i") for i in 0:(a.tentativas-1)]

    aceitos = String[]
    try
        for (i, token) in enumerate(tokens)
            CriptografiaAlvo(bytes2hex(random_bytes(16)))
            try
                r = decrypt(alvo, token)
                push!(aceitos, "tentativa $(i-1) aceitou (retornou $r)")
            catch
                # esperado
            end
        end
    finally
        ENV["FBC_KEY"] = chave_original
    end

    if !isempty(aceitos)
        return ResultadoAtaque(
            nome(a),
            true,
            "critica",
            "$(length(aceitos)) de $(a.tentativas) tokens foram aceitos com a chave errada",
            Dict{String,Any}("exemplos" => aceitos[1:min(5, length(aceitos))]),
        )
    end

    return ResultadoAtaque(
        nome(a),
        false,
        "info",
        "Nenhum dos $(a.tentativas) tokens foi aceito com chave incorreta",
    )
end

File: packages/php/src/Seguranca/Ataques/AtaqueChaveErrada.php
<?php

namespace Application\Seguranca\Ataques;

use Application\Seguranca\AlvoCriptografico;
use Application\Seguranca\AtaqueInterface;
use Application\Seguranca\CriptografiaAlvo;
use Application\Seguranca\ResultadoAtaque;

/**
 * Garante que um token só decifra com a chave correta: com qualquer outra
 * chave de 32 bytes, decrypt() tem que lançar. Verifica também que a chave
 * errada não "quase funciona" (ex: aceitar às vezes por MAC fraco).
 *
 * Cuidado: CriptografiaAlvo escreve a chave no ambiente do processo; a chave
 * original é restaurada no final pra não afetar os outros ataques.
 */
class AtaqueChaveErrada implements AtaqueInterface
{
    public function __construct(private int $tentativas = 30)
    {
    }

    public function nome(): string
    {
        return 'Rejeição de chave incorreta';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        if (!$alvo instanceof CriptografiaAlvo) {
            throw new \Application\Seguranca\SkipAtaqueException('Precisa de CriptografiaAlvo para trocar a chave.');
        }

        $chaveOriginal = $alvo->chaveDeTeste();

        $tokens = [];
        for ($i = 0; $i < $this->tentativas; $i++) {
            $tokens[] = $alvo->encrypt("MENSAGEM_SECRETA_$i");
        }

        $aceitos = [];
        try {
            foreach ($tokens as $i => $token) {
                new CriptografiaAlvo(bin2hex(random_bytes(16)));
                try {
                    $r = $alvo->decrypt($token);
                    $aceitos[] = "tentativa $i aceitou (retornou $r)";
                } catch (\Throwable $e) {
                    // esperado
                }
            }
        } finally {
            putenv('FBC_KEY=' . $chaveOriginal);
        }

        if (!empty($aceitos)) {
            return new ResultadoAtaque(
                $this->nome(),
                vulneravel: true,
                severidade: 'critica',
                detalhes: count($aceitos) . " de {$this->tentativas} tokens foram aceitos com a chave errada",
                dados: ['exemplos' => array_slice($aceitos, 0, 5)],
            );
        }

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: false,
            severidade: 'info',
            detalhes: "Nenhum dos {$this->tentativas} tokens foi aceito com chave incorreta",
        );
    }
}
